internal/controller/controllers/clusterdeployments_controller.go

```diff
@@ -101,6 +104,30 @@ func (r *ClusterDeploymentsReconciler) Reconcile(ctx context.Context, req ctrl.R
 		return ctrl.Result{}, nil
 	}
 
+	if clusterDeployment.ObjectMeta.DeletionTimestamp.IsZero() { // clusterDeployment no being deleted
+		// Register a finalizer if it is absent.
+		if !funk.ContainsString(clusterDeployment.GetFinalizers(), ClusterDeploymentFinalizerName) {
+			controllerutil.AddFinalizer(clusterDeployment, ClusterDeploymentFinalizerName)
+			if err = r.Update(ctx, clusterDeployment); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else { // clusterDeployment is being deleted
+		if funk.ContainsString(clusterDeployment.GetFinalizers(), ClusterDeploymentFinalizerName) {
+			// deletion finalizer found, deregister the backend cluster and delete agents
+			reply, cleanUpErr := r.deregisterClusterIfNeeded(ctx, req.NamespacedName)
+			if cleanUpErr != nil {
+				return reply, cleanUpErr
+			}
+
+			// remove our finalizer from the list and update it.
+			controllerutil.RemoveFinalizer(clusterDeployment, ClusterDeploymentFinalizerName)
+			if err = r.Update(ctx, clusterDeployment); err != nil {
```

Comment:
Typically a Reconcile function would return after doing this. Further actions, like trying to update status, would fail. If there's anything that happens below that's relevant to cleanup, it should probably happen in this block prior to removing the finalizer.

Reply:
Done